### ISSUE TITLE
Realm: Add recycling of sparsity maps

### ISSUE THREAD

**Issue by lightsighter** [StanfordLegion/legion#1192]
Currently Realm's sparsity maps are leaked when they are destroyed and the IDs cannot be recycled. This needs to be fixed if we're regularly going to be creating and destroying sparsity maps, e.g. in the case where we are frequently calling image/preimage for gather/scatter copy operations in Legion in order to avoid running out of sparsity map IDs. It will also guarantee that the memory associated with sparsity maps is cleaned up.

**Comment by lightsighter** [comment]
Adding @streichler and @alexaiken for discussion. I'll continue to be the owner of this issue for now since I'll likely be the one that ends up implementing it in Realm.

**Comment by apryakhin** [comment]
PR to recycle sparse maps in realm has been out for review:
- https://gitlab.com/StanfordLegion/legion/-/merge_requests/764

Depending on the priority we need to get that tested by legion team, reviewed and merged.

**Comment by apryakhin** [comment]
@lightsighter We won't be making it for 24.09, so I will have to happen in the next release.

**Comment by lightsighter** [comment]
Yes, I agree it's not going to make this release. We'll get it in the next one.

**Comment by apryakhin** [comment]
@reazulhoque

**Comment by apryakhin** [comment]
@lightsighter Just to follow up on this. We have the branch with realm changes that addresses the out-of-order reference counting. I think there is still some work on the Legion side that needs to happen. @reazulhoque  for more context. It may so happen that legion priorities have shifted for the upcoming release?  If yes just make sure we have it here in wording and have a consensus on when we want this to be done.

**Comment by lightsighter** [comment]
I think we still want to get this merged by the end of the year. Probably not before the December release but right after it. @reazulhoque has something more important that he's doing now, but coming back to clean this up will be his next priority.
